app/src/rgb_underglow.c

```diff
@@ -292,101 +265,144 @@ int zmk_rgb_underglow_save_state() {
 #endif
 }
 
-int zmk_rgb_underglow_cycle_effect(int direction) {
+int zmk_rgb_underglow_get_state(bool *on_off) {
     if (!led_strip)
         return -ENODEV;
 
-    if (state.current_effect == 0 && direction < 0) {
-        state.current_effect = UNDERGLOW_EFFECT_NUMBER - 1;
-        return 0;
-    }
+    *on_off = state.on;
+    return 0;
+}
 
-    state.current_effect += direction;
+int zmk_rgb_underglow_on() {
+    if (!led_strip)
+        return -ENODEV;
 
-    if (state.current_effect >= UNDERGLOW_EFFECT_NUMBER) {
-        state.current_effect = 0;
+#if IS_ENABLED(CONFIG_ZMK_RGB_UNDERGLOW_EXT_POWER)
+    if (ext_power != NULL) {
+        int rc = ext_power_enable(ext_power);
+        if (rc != 0) {
+            LOG_ERR("Unable to enable EXT_POWER: %d", rc);
+        }
     }
+#endif
 
+    state.on = true;
     state.animation_step = 0;
+    k_timer_start(&underglow_tick, K_NO_WAIT, K_MSEC(50));
 
     return zmk_rgb_underglow_save_state();
 }
 
-int zmk_rgb_underglow_toggle() {
+int zmk_rgb_underglow_off() {
     if (!led_strip)
         return -ENODEV;
 
-    state.on = !state.on;
-
 #if IS_ENABLED(CONFIG_ZMK_RGB_UNDERGLOW_EXT_POWER)
     if (ext_power != NULL) {
-        int rc;
-
-        if (state.on) {
-            rc = ext_power_enable(ext_power);
-        } else {
-            rc = ext_power_disable(ext_power);
-        }
-
+        int rc = ext_power_disable(ext_power);
         if (rc != 0) {
-            LOG_ERR("Unable to toggle EXT_POWER: %d", rc);
+            LOG_ERR("Unable to disable EXT_POWER: %d", rc);
         }
     }
 #endif
 
-    if (state.on) {
-        state.animation_step = 0;
-        k_timer_start(&underglow_tick, K_NO_WAIT, K_MSEC(50));
-    } else {
-        zmk_rgb_underglow_off();
+    for (int i = 0; i < STRIP_NUM_PIXELS; i++) {
+        pixels[i] = (struct led_rgb){r : 0, g : 0, b : 0};
+    }
+
+    led_strip_update_rgb(led_strip, pixels, STRIP_NUM_PIXELS);
+
+    k_timer_stop(&underglow_tick);
+    state.on = false;
 
-        k_timer_stop(&underglow_tick);
+    return zmk_rgb_underglow_save_state();
+}
+
+int zmk_rgb_underglow_cycle_effect(int direction) {
+    if (!led_strip)
+        return -ENODEV;
+
+    if (state.current_effect == 0 && direction < 0) {
+        state.current_effect = UNDERGLOW_EFFECT_NUMBER - 1;
+        return 0;
     }
 
+    state.current_effect += direction;
+
+    if (state.current_effect >= UNDERGLOW_EFFECT_NUMBER) {
+        state.current_effect = 0;
+    }
+
+    state.animation_step = 0;
```

Comment:
```suggestion
    state.current_effect += direction % UNDERGLOW_EFFECT_NUMBER;

    if (state.current_effect < 0) {
        state.current_effect += UNDERGLOW_EFFECT_NUMBER;
    }

    state.animation_step = 0;
```
I think this simplifies the logic and would support the possibility of a different direction that -1 or 1 (if someone wants that for some reason). Also the return statement on line 327 probably should be `zmk_rgb_underglow_save_state()` if we don't switch to this.

Reply:
Used a similar modulo trick as in the other areas instead. Just pushed the updated version.